### ISSUE TITLE
How to make Dropzone more correct?

### ISSUE THREAD

**Issue by resistorsoftware** [baoagency/polaris_view_components#217]
Home controller renders Polaris View Components in index.html.erb
Dropzone renders as a component. Dropzone is a form. Form is for Direct Upload. User picks a file, clicks submit. Form Submit starts file uploading. So a POST to the home controller. 

    <form enctype="multipart/form-data" action="/home?shop=cold-coke.myshopify.com" accept-charset="UTF-8" method="post">

The parameters I see tell me quite the story. 

    Started POST "/rails/active_storage/direct_uploads?shop=cold-coke.myshopify.com" for 127.0.0.1 at 2022-04-04 20:31:19 -0400
    20:31:19 web.1    | Processing by ActiveStorage::DirectUploadsController#create as JSON
    20:31:19 web.1    |   Parameters: {"blob"=>{"filename"=>"cokes.csv", "content_type"=>"text/csv", "byte_size"=>1963, "checksum"=>"yXHzeZyoEvajBc3ugai9qQ=="}, "direct_upload_token"=>"[FILTERED]", "attachment_name"=>nil, "shop"=>"cold-cokes.myshopify.com", "host"=>"aG90d2lyZS1yZXNpc3Rvci5teXNob3BpZnkuY29tL2FkbWlu", "direct_upload"=>{"blob"=>{"filename"=>"cokes.csv", "content_type"=>"text/csv", "byte_size"=>1963, "checksum"=>"yXHzeZyoEvajBc3ugai9qQ=="}, "direct_upload_token"=>"[FILTERED]", "attachment_name"=>nil}}

So Rails chugs away and does stuff. The see ActiveStorage do a PUT and my data thus lives in the cloud. The form submit was hijacked by Turbo, and so I see this POST is a Turbo Stream request. Do I care? Is that good, bad or ugly? At this point I would like to think that I could escape unharmed!

    Started POST "/home?shop=cold-coke.myshopify.com"
    Processing by HomeController#create as TURBO_STREAM
    Parameters: {"authenticity_token"=>"[FILTERED]", "upload_file"=>"eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBjZz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--95ac7406a28fa890c6bf40d081e6c5d8204f62e1", "shop"=>"cold-coke.myshopify.com", "host"=>"aG90d2lyZS1yZXNpc3Rvci5teXNob3BpZnkuY29tL2FkbWlu"}

This home controller create can thus attach the file provided, and we can run a sidekiq job to process the file. Yay! And then I have no choice but to do a Turbo Redirect. But that is not nice, as it redirects me to the index of home controller, and that re-runs a lot of SQL and other intense stuff I would rather not re-render. So how to not do this:

     @message = "Uploaded your file. It is being processed now!"
     turbo_redirect_to home_path, notice: @message
     
This sort of works! One small problem with it, in Safari browser, that message is not rendered. Firefox works fine though.

TLDR: How can I deal with this POST, and not re-render the home page? Should I respond in the create with a turbo stream response? But then that would imply when the homepage renders, it renders the dropzone with a partial? And then I do that? But if I do that, I don't get a Flash Message. So what to do there? Can I respond with a small partial, and a Flash? 

I think I am getting this... but there seem to be a lot of little niggling details.

**Comment by forsbergplustwo** [comment]
Hey @resistorsoftware,

This is not really an "issue" related to this repo, so maybe better posted on Stack Overflow in the future?

>  Should I respond in the create with a turbo stream response? But then that would imply when the homepage renders, it renders the dropzone with a partial? And then I do that? But if I do that, I don't get a Flash Message. So what to do there?

Yes, that would be the way I'd do it. Use a turbostream responce, which re-renders the upload form, and also adds any flash message at the same time. You can do both in a single file:

```
<%= turbo_stream.replace @setting, partial: "form", locals: { setting: @setting } %>
<%= turbo_stream.replace "flash-id" partial: "layouts/flash" %>
```
Take a look at: 

https://github.com/kirillplatonov/shopify-hotwire-sample/blob/main/app/views/layouts/_flash_messages.turbo_stream.erb

and

https://github.com/kirillplatonov/shopify-hotwire-sample/blob/main/app/javascript/controllers/flash_controller.js

The idea is that you re-insert the flash div into the html, which includes a Stimulus Controller, which then checks and displays any new message in it's Connect method.

Hope it helps!

**Comment by resistorsoftware** [comment]
I mention it here in this Repo, because this Repo does not exist in a vacuum! Those that choose to use these components are using them without any "usage" examples. So I am pointing out that those that choose to use the DropZone as an example, might need to think about how it is actually used. Going on StackOverflow, while it sounds appealing, is probably as far from a solution as one could get. No one there will understand a thing about **this** project.

If you are suggesting this Repo contains nothing in issues but bare naked Polaris View Component structures, with zero about how to effectively use it, fine. **I cannot argue**. I beg to differ though, and feel like what this project really needs is people using it and talking about it. 

I have struggled to learn Rails through this project and all its nutty stuff, and now that I am close to having it all work, I am excited. I am still being hammered though by things not working as perhaps they should. Is it me? It is memorex?

**Comment by resistorsoftware** [comment]
I appreciate the existence of this Repo. Don't get me wrong. It has been a neat exchange of doing it all React Polaris to Rails ... that being said, it takes some time to adjust not only to Rails 7 but to how these components work. 

As an example, if you use Dropzone, and someone initiates an upload, and they try move off-page, there is supposed to be an ALERT triggered in the browser that tells them "Hey! Do you really want to cancel this Upload???". In order for me to add that, assuming it is not in this repo, I have to initiate a PR. OK fine. I get that. But I like to first talk it out. Get a feel from the crowd about whether I am barking up the wrong tree or not. 

I don't want to just hack my version of this code!

**Comment by kirillplatonov** [comment]
It seems like the question is really about how Hotwire Turbo and specifically forms & turbo streams are working. This is really outside of the scope of Polaris View Components. Hotwire has active community where you can get help with questions like this: https://discuss.hotwired.dev
